### PR TITLE
Provide cmetrics API

### PR DESCRIPTION
**Pull request by cosmo0920**

### CODE REVIEW

**Comment by niedbalski** [comment]
@cosmo0920 2 observations.

1. Please adjust the plugin to use the plugin.go interface.
2. Lets pass a third parameter (metrics) to the plugin initialisation function, so the plugin itself doesn't have to initialise that context by itself (https://github.com/calyptia/plugin/blob/main/cshared.go#L74 pass it as part of the ctx directly or as a third parameter.)

File: testdata/Dockerfile [hunk 1]
```diff
@@ -10,10 +19,14 @@ RUN go mod verify
...
 
+ADD https://github.com/calyptia/cmetrics/releases/download/${CMETRICS_RELEASE}/cmetrics_${CMETRICS_VERSION}_${PACKAGEARCH}-headers.deb external/
+ADD https://github.com/calyptia/cmetrics/releases/download/${CMETRICS_RELEASE}/cmetrics_${CMETRICS_VERSION}_${PACKAGEARCH}.deb external/
+RUN dpkg -i external/*.deb
+
 RUN go build -trimpath -buildmode c-shared -o ./bin/go-test-input-plugin.so ./testdata/input/input.go
 RUN go build -trimpath -buildmode c-shared -o ./bin/go-test-output-plugin.so ./testdata/output/output.go
 
-FROM ghcr.io/calyptia/calyptia-fluent-bit/advanced:main
+FROM ghcr.io/calyptia/enterprise/advanced:pr-123
```
Comment (niedbalski) [inline testdata/Dockerfile:29]: we should remove this before merging.

Comment (cosmo0920) [reply testdata/Dockerfile:29]: Yes, just for testing purpose. I'm waiting for https://github.com/calyptia/calyptia-fluent-bit/pull/123 patch and published advanced image.

**Review by niedbalski — CHANGES_REQUESTED** [review]
@cosmo0920 can you place an example in the README.md and rename the image to point to https://github.com/calyptia/calyptia-fluent-bit/pkgs/container/enterprise%2Fadvanced/23935905?tag=main?

File: examples/in_gdummy/in_gdummy.go [hunk 2]
```diff
@@ -1,65 +1,70 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			if err != nil && !errors.Is(err, context.Canceled) {
+				err = counter_failure.Inc(time.Now(), []string{"in_gdummy"})
+				if err != nil {
+					return err
+				}
+				return err
```
Comment (gabrielbussolo) [inline examples/in_gdummy/in_gdummy.go:51]: if u will return the `err` anyway so maybe u can do this:
```suggestion
				return counter_failure.Inc(time.Now(), []string{"in_gdummy"})
```

**Review by gabrielbussolo — COMMENTED** [review]
i dont have whole context on the subject, so dont will aprove, just comment :)

File: examples/in_gdummy/in_gdummy.go (hunk 2 above)
Comment (cosmo0920) [reply examples/in_gdummy/in_gdummy.go:51]: Ah, it seems good!

File: plugin.go [hunk 3]
```diff
@@ -35,13 +37,13 @@ func init() {
 
 // InputPlugin interface to represent an input fluent-bit plugin.
 type InputPlugin interface {
-	Init(ctx context.Context, conf ConfigLoader) error
+	Init(ctx context.Context, conf ConfigLoader, cmt *cmetrics.Context) error
```
Comment (nicolasparada) [inline plugin.go:40]: Is this somethig the plugin author should use?
If we want to have metrics about a plugin failing or its latency, we can do it internally in the plugin package itself.

And in the case we want to expose these internals, they should be an interface.

But I strongly suggest not exposing it.

Comment (niedbalski) [reply plugin.go:40]: > Is this somethig the plugin author should use? If we want to have metrics about a plugin failing or its latency, we can do it internally in the plugin package itself.
> 
> And in the case we want to expose these internals, they should be an interface.
> 
> But I strongly suggest not exposing it.

Yes, the plugin author should explicitly be able declare arbitrary metrics, and should be extensible enough. What's your idea around this?

Comment (niedbalski) [reply plugin.go:40]: @nicolasparada are you thinking to compose a MetricsHandler interface?

Comment (nicolasparada) [reply plugin.go:40]: But yeah, if we want to expose the functionallity it should be through an interface.

> Accept interfaces, return structs.

We might have something like...
```go
type Counter interface {
  Add(delta float64)
}

type Gauge interface {
  Add(delta float64)
  Set(value float64)
}

type Histogram interface {
  Observe(value float64)
}

type Metrics interface {
  NewCounter(name string) Counter
  NewGauge(name string) Gauge
  NewHistogram(name string) Histogram
}
```

So this becomes quite easy for the plugin author.

```go
type myPlugin struct {
  myCounter plugin.Counter
}

func (plug *thePlugin) Init(ctx context.Context, conf plugin.ConfigLoader, metrics plugin.Metrics) error {
	plug.myCounter = metrics.NewCounter("my_counter")
	return nil
}

func (plug thePlugin) Collect(ctx context.Context, ch chan<- plugin.Message) error {
}

Comment (niedbalski) [reply plugin.go:40]: Yes, @nicolasparada can you implement this? , I will dismiss #16 though, as we will follow the default metrics using the output plugins approach to be consistent.

Comment (nicolasparada) [reply plugin.go:40]: Yeah, I think I can. SHall I do it on top of this PR?

Comment (niedbalski) [reply plugin.go:40]: Sure @nicolasparada

**Review by niedbalski — APPROVED** [review]
LGTM

**Pull request merged**